configuration/src/main/java/io/camunda/configuration/UnifiedConfigurationHelper.java

```diff
@@ -0,0 +1,253 @@
+
+  public UnifiedConfigurationHelper(@Autowired Environment environment) {
+    // We need to pin the environment object statically so that it can be used to perform the
+    // fallback mechanism.
+    UnifiedConfigurationHelper.environment = environment;
+  }
+
+  public static <T> T validateLegacyConfiguration(
+      final String newProperty,
+      final T newValue,
+      final Class<T> expectedType,
+      final boolean legacyConfigAllowed,
+      final boolean onlyIfValuesMatch) {
+
+    // Input validation
+    if (!legacyConfigAllowed && onlyIfValuesMatch) {
+      throw new UnifiedConfigurationException(
+          "onlyIfValuesMatch cannot be true without legacyConfigAllowed.");
+    }
+
+    // Tools
+    final Set<String> legacyProperties = legacyPropertiesDict.get(newProperty);
+
+    // Backwards compatibility not supported
+    if (!legacyConfigAllowed) {
+      return backwardsCompatibilityNotSupported(legacyProperties, newProperty, newValue);
+    }
+
+    // Backwards compatibility supported  only if values match
+    if (legacyConfigAllowed && onlyIfValuesMatch) {
+      final T legacyValue = getLegacyValue(legacyProperties, expectedType);
+      return backwardsCompatibilitySupportedOnlyIfValuesMatch(
+          legacyProperties, legacyValue, newProperty, newValue);
+    }
+
+    // Backwards compatibility supported
+    final T legacyValue = getLegacyValue(legacyProperties, expectedType);
+    return backwardsCompatibilitySupported(legacyProperties, legacyValue, newProperty, newValue);
+  }
+
+  private static <T> T getLegacyValue(
+      final Set<String> legacyProperties, final Class<T> expectedType) {
+    final Set<T> legacyValues = new HashSet<>();
+
+    for (final String legacyProperty : legacyProperties) {
+      final String strValue = environment.getProperty(legacyProperty);
+      final T legacyValue = parseLegacyValue(strValue, expectedType);
+      legacyValues.add(legacyValue);
+    }
+
+    if (legacyValues.size() > 1) {
+      throw new UnifiedConfigurationException(
+          String.format(
+              "Ambiguous legacy configuration. Legacy properties: %s; Legacy values: %s",
+              String.join(", ", legacyProperties), legacyValues.toString()));
+    }
+
+    return legacyValues.iterator().next();
+  }
+
+  private static <T> T backwardsCompatibilityNotSupported(
+      final Set<String> legacyProperties, final String newProperty, final T newValue) {
+    final boolean legacyPresent = legacyConfigPresent(legacyProperties);
+    final boolean newPresent = newConfigPresent(newProperty);
+
+    if (!legacyPresent && !newPresent) {
+      // Legacy config: not present
+      // New config...: not present
+      // We can return the default value
+      return newValue;
+    } else if (!legacyPresent && newPresent) {
+      // Legacy config: not present
+      // New config...: present
+      // We can return newValue, that's either the default or the configured one
+      return newValue;
```

Comment:
These two cases can be combined.
```suggestion
    if (!legacyPresent ) {
      // Legacy config: not present
      // We can return the newValue or default value
      return newValue;
```